model/src/gateway/event/gateway.rs

```diff
@@ -28,9 +30,83 @@ enum Field {
     T,
 }
 
-struct GatewayEventVisitor;
+pub struct GatewayEventDeserializer<'a> {
+    event_type: Option<&'a str>,
+    op: u8,
+}
+
+impl<'a> GatewayEventDeserializer<'a> {
+    // Create a gateway event deserializer with some information found by
+    // scanning the JSON payload to deserialise.
+    //
+    // This will scan the payload for the opcode and, optionally, event type if
+    // provided. The opcode key ("op"), must be in the payload while the event
+    // type key ("t") is optional and only required for event ops.
+    pub fn new(input: &'a str) -> Option<Self> {
+        let op = Self::find_opcode(input)?;
+        let event_type = Self::find_event_type(input);
+
+        Some(Self { event_type, op })
+    }
+
+    fn find_event_type(input: &'a str) -> Option<&'a str> {
+        // Discord seems to usually (always?) provide the type at the end of the
+        // payload, so let's search from the right...
+        let from = input.rfind(r#""t":"#)? + 4;
+
+        // Now let's find where the value starts. There may or may not be any
+        // amount of whitespace after the "t" key.
+        let start = input.get(from..)?.find('"')? + from + 1;
+        let to = input.get(start..)?.find('"')?;
+
+        input.get(start..start + to)
+    }
+
+    fn find_opcode(input: &'a str) -> Option<u8> {
+        // Find the op key's position and then search for where the first
+        // character that's not base 10 is. This'll give us the bytes with the
+        // op which can be parsed.
+        //
+        // Add on 5 at the end since that's the length of what we're finding.
+        let from = input.find(r#""op":"#)? + 5;
+
+        // Look for the first thing that isn't a base 10 digit or whitespace,
+        // i.e. a comma (denoting another JSON field), curly brace (end of the
+        // object), etc. This'll give us the op number, maybe with a little
+        // whitespace.
+        let to = input.get(from..)?.find(|c: char| c == ',' || c == '}')?;
```

Comment:
```suggestion
        let to = input.get(from..)?.find(&[',', '}'] as &[_])?;
```
unsure if this is faster, it would mean that a closure is not called. It may be in-lined though.

Reply:
Looking at GodBolt, it looks like the second is inlined more aggressively.